deepspeed/runtime/domino/transformer.py

```diff
@@ -356,20 +382,73 @@ def forward(self, hidden_states, attention_mask, rotary_pos_emb=None):
         if not self.llama_model:
             rotary_pos_emb = None
 
-        attention_output0, attention_bias0  = \
-            self.self_attention(
-                layernorm_output0,
+        if self.input_split_dim == "seq":
+            # Compute full Q, K, V
+            layernorm_output = torch.concat([layernorm_output0, layernorm_output1], dim=0)
+            mixed_x_layer, _ = self.query_key_value(layernorm_output)
+
+            # [s, b, np * 3 * hn] --> [s, b, np, 3 * hn]
+            new_tensor_shape = mixed_x_layer.size()[:-1] + (
+                self.num_attention_heads_per_partition,
+                3 * self.hidden_size_per_attention_head,
+            )
+            mixed_x_layer = mixed_x_layer.view(*new_tensor_shape)
+            
+            # [s, b, np, 3 * hn] -> [b, np, s, 3*hn]
+            mixed_x_layer = mixed_x_layer.permute(1, 2, 0, 3).contiguous()
+            # [s, b, np, 3 * hn] --> [s, b, np, hn], [s, b, np, hn], [s, b, np, hn]
+            (query_layer, key_layer, value_layer) = torch.split(mixed_x_layer, [
+                self.hidden_size_per_attention_head, self.hidden_size_per_attention_head,
+                self.hidden_size_per_attention_head
+            ], dim=3)
+
+            # [s, b, np, np * hn] -> [s, b, np, hn]
+            query_layer = query_layer.view(query_layer.size(0), query_layer.size(1), -1,
+                                        self.hidden_size_per_attention_head)
+            
+            if rotary_pos_emb is not None:
+                if isinstance(rotary_pos_emb, tuple):
+                    rotary_pos_emb = rotary_pos_emb
+                else:
+                    rotary_pos_emb = ((rotary_pos_emb, ) * 2)
+                q_pos_emb, k_pos_emb = rotary_pos_emb
+                query_layer = self.apply_rotary_pos_emb(query_layer, q_pos_emb)
+                key_layer = self.apply_rotary_pos_emb(key_layer, k_pos_emb)
+
+            batchsize, num_heads, seq_len, hidden_per_head = query_layer.shape[0], query_layer.shape[1], query_layer.shape[2], query_layer.shape[3]
+
+            # seq 0: core attention
+            context_layer0 = self.self_attention_sp(query_layer[:, :, :seq_len//2, :], key_layer, value_layer, attention_mask[:, :, :seq_len//2, :])
+            # Output. [s, b, h]
+            attention_output0, attention_bias0 = self.dense(context_layer0)
+
+            handle0 = dist.all_reduce(attention_output0, group=self.mpu.get_tensor_model_parallel_group(), async_op=True)
+
+            # seq 1: core attention
+            context_layer1 = self.self_attention_sp(query_layer[:, :, seq_len//2:, :], key_layer, value_layer, attention_mask[:, :, seq_len//2:, :])
+            # Output. [s, b, h]
+            attention_output1, attention_bias1 = self.dense(context_layer1)
+
+            handle1 = dist.all_reduce(attention_output1, group=self.mpu.get_tensor_model_parallel_group(), async_op=True)
+            handle0.wait()
```

Comment:
these are exactly the same as sharedAttention forward, I don't see why we need these duplication here. 

Also please follow our current code hierarchy, not pull up lower layer module implementation code to upper layer module. 

e.g., if any real change need to make on `sharedAttention`, create a similar module say `XYZAttention`, then here in `DominoTransformerLayer` forward function, we can simply call `XYZAttention` module without duplicating XYZAttention module's every line of code of its forward.